packages/entity-server/src/routes/hierarchy/middleware/fields.ts

```diff
@@ -0,0 +1,119 @@
+  ...validFields,
+  ...Object.keys(extendedFieldFunctions),
+] as (keyof ExtendedEntityFields)[];
+
+export const extractFieldsFromQuery = (queryFields?: string) => {
+  if (!queryFields) {
+    return allFields; // Display all fields if none specifically requested
+  }
+
+  const requestedFields = queryFields.split(',');
+  const fields = new Set<keyof ExtendedEntityFields>();
+  requestedFields.forEach(requestedField => {
+    if (validateField(requestedField)) {
+      fields.add(requestedField);
+    } else {
+      throw new Error(`Unknown field requested: ${requestedField}`);
+    }
+  });
+  return Array.from(fields);
+};
+
+type Writable<T> = { -readonly [field in keyof T]?: T[field] };
+export const mapEntityToFields = (fields: (keyof ExtendedEntityFields)[]) => async (
+  entity: EntityType,
+  context: HierarchyContext,
+) => {
+  const mappedEntity: Writable<EntityResponseObject> = {};
+  for (let i = 0; i < fields.length; i++) {
+    const field = fields[i];
+    if (isEntityField(field)) {
+      mappedEntity[field] = entity[field];
+    } else {
+      (mappedEntity[field] as string | string[] | undefined) = await extendedFieldFunctions[field](
+        entity,
+        context,
+      );
+    }
+  }
+  return mappedEntity;
+};
+
+export const mapEntitiesToFields = (
+  req: HierarchyRequest,
+  fields: (keyof ExtendedEntityFields)[],
+) => async (entities: EntityType[], context: HierarchyContext) => {
+  const relationRecords =
+    fields.includes('parent_code') || fields.includes('children_codes')
+      ? await req.models.ancestorDescendantRelation.getImmediateRelations(context.hierarchyId)
+      : [];
+  const mappedEntities: Writable<EntityResponseObject>[] = new Array(entities.length)
+    .fill(0)
+    .map(() => ({})); // fill array with empty objects
+  for (let i = 0; i < fields.length; i++) {
+    const field = fields[i];
+    if (isEntityField(field)) {
+      entities.forEach((entity, index) => {
+        mappedEntities[index][field] = entity[field];
+      });
+    } else {
+      switch (field) {
+        case 'parent_code': {
+          const childToParentMap = reduceToDictionary(
+            relationRecords,
+            'descendant_code',
+            'ancestor_code',
+          );
+          entities.forEach((entity, index) => {
+            mappedEntities[index].parent_code = getParentCodeBulk(entity, childToParentMap);
+          });
+          break;
+        }
+        case 'children_codes': {
+          const parentToChildrenMap = reduceToArrayDictionary(
+            relationRecords,
+            'ancestor_code',
+            'descendant_code',
+          );
+          entities.forEach((entity, index) => {
+            mappedEntities[index].children_codes = getChildrenCodesBulk(
+              entity,
+              parentToChildrenMap,
+            );
+          });
+          break;
+        }
+        default: {
+          throw new Error(`Unknown extended entity field requested: ${field}`);
```

Comment:
Ideally we won't be throwing an error here but instead just blindly calling `await extendedFieldFunctions[field](entity, context);` like we do in `mapEntityToFields()`. However Typescript is being a little too smart here and knows that if the field isn't either `parent_code` or `children_codes` then it's invalid and forces us to throw an error.

If we add more `extendedFieldFunctions` in the future that doesn't require any special logic when performing on multiple entities (region data, etc.) then we can replace throw statement with the behaviour above 👍